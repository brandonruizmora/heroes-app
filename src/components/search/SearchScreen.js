import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../heroes/HeroCard';
import { getHeroesByName } from '../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {

    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);


    const initialForm = {
        hero: q,
    }

    const [formValues, handleInputChange] = useForm(initialForm);

    const { hero } = formValues;

    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

    const handleSubmit = (e) => {
        e.preventDefault();
        history.push(`?q=${hero}`);
    }

    return (
        <div>
            <h1>Search Screen</h1>
            <hr />
            <div className="row">
                <div className="col-5">
                    <h4>Search Form</h4>
                    <hr />
                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="hero"
                            value={hero}
                            onChange={handleInputChange}
                            className="form-control mb-3"
                            placeholder="Find your hero"
                            autoComplete="off"
                        />
                        <div class="d-grid gap-2">
                            <button
                                type="submit"
                                className="btn btn-outline-success"
                            >
                                Search
                            </button>
                        </div>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr />
                    {
                        (q === '') &&
                        <div className="alert alert-info" role="alert">
                            Search a hero...
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                        <div className="alert alert-danger" role="alert">
                            There is any hero with {hero}
                        </div>
                    }
                    {
                        heroesFiltered.map(hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </div>
    )
}
