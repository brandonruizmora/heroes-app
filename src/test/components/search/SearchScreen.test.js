import '../../../setupTests';
import { shallow, mount } from 'enzyme';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter } from 'react-router';

describe('Pruebas en <SearchScreen />', () => {

    const history = {
        push: jest.fn()
    }

    const wrapper = mount(
        <MemoryRouter>
            <SearchScreen history={history} />

        </MemoryRouter>
    )

    test('Debe de renderizarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero...');

    });

    test('Debe de mostrar a batman y el valor del input en el query string', () => {
        const history = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen history={history} />

            </MemoryRouter>
        )
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('input').prop('value')).toBe('batman');
    });

    test('Debe de mostrar error si no se encuentra el hero', () => {
        const history = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen history={history} />

            </MemoryRouter>
        );
        expect(wrapper.find('.alert-danger').exists()).toBe(true);
    });

    test('Debe de llamar el push del history', () => {
        const history = {
            push: jest.fn()
        }

        const wrapper = mount(
            <MemoryRouter>
                <SearchScreen history={history} />

            </MemoryRouter>
        ); 

        wrapper.find('input').simulate('change', {
            target: {
                name: 'hero',
                value: 'batman'
            }
        });
        wrapper.find('form').simulate('submit');

        expect(history.push).toHaveBeenCalledWith('?q=batman');


    });
    


});
