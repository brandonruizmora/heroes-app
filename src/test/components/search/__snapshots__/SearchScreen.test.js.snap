// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> Debe de mostrar a batman y el valor del input en el query string 1`] = `
<div
  className="container"
>
  <h1
    className="fw-bold text-shadow-white"
  >
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-12 col-sm-5 bg-white rounded p-5 h-25"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-3"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-success"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-12 col-sm-7"
    >
      <h4
        className="text-shadow-white"
      >
        Results
      </h4>
      <hr />
      <div
        className="card m-3"
      >
        <div
          className="row g-0"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="dc-batman"
              className="img-fluid rounded-start"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              />
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                MÃ¡s
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> Debe de renderizarse correctamente 1`] = `
<div
  className="container"
>
  <h1
    className="fw-bold text-shadow-white"
  >
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-12 col-sm-5 bg-white rounded p-5 h-25"
    >
      <h4>
        Search Form
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control mb-3"
          name="hero"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <div
          className="d-grid gap-2"
        >
          <button
            className="btn btn-outline-success"
            type="submit"
          >
            Search
          </button>
        </div>
      </form>
    </div>
    <div
      className="col-12 col-sm-7"
    >
      <h4
        className="text-shadow-white"
      >
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
        role="alert"
      >
        Search a hero...
      </div>
    </div>
  </div>
</div>
`;
